### ISSUE TITLE
Poissible problems with BatchNorm

### ISSUE THREAD

**Issue by vodp** [ethereon/caffe-tensorflow#40]
I trained GoogleNet (v1) with BatchNorm on Caffe. Prediction is OK on Caffe. When converting to TF model, prediction results are quite random and unpredictable! Then I trained another GoogleNet (v1) without BatchNorm layers on Caffe. Now predictions are OK on both Caffe and the converted TF model. So I think there may be some bugs relevant to BatchNorm part. Have you tested it?

Best,

**Comment by ethereon** [comment]
I've tested the BatchNorm conversion on ResNet, which appeared to work OK (accuracy numbers listed in the readme). That said, I wouldn't be surprised if there's a bug in there.

I'll take another look at it. Would it be possible for you to share the problematic GoogleNet model?

**Comment by loweew** [comment]
I'm also experiencing this. Any updates or workarounds?

**Comment by ethereon** [comment]
So far, my tests indicate that the batch norm conversion works correctly.

In addition to the official ResNet models, I recently also tested it on models trained from scratch in Caffe. In all cases, the results are near identical (up to 6 decimal places).

**Comment by ethereon** [comment]
Closing this since I haven't been able to reproduce this. Feel free to re-open if you encounter an issue with batch norm.

**Comment by jzhaosc** [comment]
A BatchNorm layer in tensorflow is equivalent to two successive layers in caffe (batchNorm + Sclale), I used the code below to successfully convert BatchNorm in TF to caffe:
 
            net.params[bn_name][0].data[:] = tf_movingmean
            # epsilon 0.001 is the default value used by tf.contrib.layers.batch_norm!!!!!
            net.params[bn_name][1].data[:] = tf_movingvariance + 0.001
            net.params[bn_name][2].data[:] = 1
            net.params[scale_name][0].data[:] = tf_gamma
            net.params[scale_name][1].data[:] = tf_beta
